File: app/src/com/ternaryop/phototumblrshare/activity/ImageViewerActivity.java
package com.ternaryop.phototumblrshare.activity;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

import android.annotation.SuppressLint;
import android.app.WallpaperManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.text.Html;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.ternaryop.phototumblrshare.AppSupport;
import com.ternaryop.phototumblrshare.R;
import com.ternaryop.tumblr.TumblrPhotoPost;
import com.ternaryop.utils.AbsProgressBarAsyncTask;
import com.ternaryop.utils.DialogUtils;
import com.ternaryop.utils.IOUtils;
import com.ternaryop.utils.ImageUtils;
import com.ternaryop.utils.ShareUtils;

@SuppressLint("SetJavaScriptEnabled")
public class ImageViewerActivity extends PhotoTumblrActivity {
	private static final String IMAGE_URL = "imageUrl";
	private static final String IMAGE_TITLE = "imageTitle";
	private static final String IMAGE_TAG = "imageTag";
	private boolean webViewLoaded;
	private Menu optionsMenu;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_webview);
        setTitle(R.string.image_viewer_activity_title);
	    setActionBarIcon();

	    ProgressBar progressBar = (ProgressBar) findViewById(R.id.webview_progressbar);
	    
	    Bundle bundle = getIntent().getExtras();
		String imageUrl = bundle.getString(IMAGE_URL);
    	String data = "<body><img src=\"" + imageUrl +"\"/></body>";
    	prepareWebView(progressBar).loadData(data, "text/html", "UTF-8");
	    try {
			getActionBar().setSubtitle(new URI(imageUrl).getHost());
		} catch (URISyntaxException e) {
		}
	}

	private WebView prepareWebView(final ProgressBar progressBar) {
		WebView webView = (WebView) findViewById(R.id.webview_view);
		webView.getSettings().setJavaScriptEnabled(true);
		webView.addJavascriptInterface(this, "dimRetriever");
		webViewLoaded = false;

		webView.setInitialScale(1);
		webView.getSettings().setLoadWithOverviewMode(true);
		webView.getSettings().setUseWideViewPort(true);		
		webView.getSettings().setBuiltInZoomControls(true);
		webView.getSettings().setSupportZoom(true);

        webView.setWebChromeClient(new WebChromeClient() {
			@Override
			public void onProgressChanged(WebView view, int newProgress) {
				progressBar.setProgress(newProgress);
			}
		});
		webView.setWebViewClient(new WebViewClient() {
			@Override
			public void onPageStarted(WebView view, String url, Bitmap favicon) {
				progressBar.setProgress(0);
				progressBar.setVisibility(View.VISIBLE);
			}
			@Override
			public void onPageFinished(WebView view, String url) {
				webViewLoaded = true;
				progressBar.setVisibility(View.GONE);
				view.loadUrl("javascript:var img = document.querySelector('img');dimRetriever.setDimensions(img.width, img.height)");
				// onPrepareOptionsMenu should be called after onPageFinished
				if (optionsMenu != null) {
					optionsMenu.findItem(R.id.action_image_viewer_wallpaper).setVisible(true);
					optionsMenu.findItem(R.id.action_image_viewer_share).setVisible(true);
				}
			}
		});
		return webView;
	}

	public static void startImageViewer(Context context, String url, TumblrPhotoPost post) {
		Intent intent = new Intent(context, ImageViewerActivity.class);
		Bundle bundle = new Bundle();

		bundle.putString(IMAGE_URL, url);
		if (post != null) {
			bundle.putString(IMAGE_TITLE, post.getCaption());
			if (!post.getTags().isEmpty()) {
				bundle.putString(IMAGE_TAG, post.getTags().get(0));
			}
		}
		intent.putExtras(bundle);

		context.startActivity(intent);
	}

	@JavascriptInterface
	public void setDimensions(final int w, final int h) {
		runOnUiThread(new Runnable() {
		    public void run() {
		        getActionBar().setSubtitle(getActionBar().getSubtitle() + String.format(" (%1dx%2d)", w, h));
			}
		  });
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.image_viewer, menu);
		return true;
	}
	
	@Override
	public boolean onPrepareOptionsMenu(Menu menu) {
		optionsMenu = menu;
		if (!webViewLoaded) {
			menu.findItem(R.id.action_image_viewer_wallpaper).setVisible(false);
			menu.findItem(R.id.action_image_viewer_share).setVisible(false);
		}
		return true;
	}
	
	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
		case R.id.action_image_viewer_wallpaper:
			setWallpaper();
			return true;
		case R.id.action_image_viewer_share:
			shareImage();
			return true;
		default:
			return super.onOptionsItemSelected(item);
		}
	}

	private void shareImage() {
		try {
			final String imageUrl = getIntent().getExtras().getString(IMAGE_URL);
			String fileName = new URI(imageUrl).getPath();
			int index = fileName.lastIndexOf('/');
			if (index != -1) {
				fileName = fileName.substring(index + 1);
			}
			// write to a public location otherwise the called app can't access to file
			final File destFile = new File(AppSupport.getPicturesDirectory(), fileName);
			new AbsProgressBarAsyncTask<Void, Void, Void>(this, getString(R.string.downloading_image)) {

				@Override
				protected Void doInBackground(Void... voidParams) {
					FileOutputStream os = null;
					InputStream is = null;
			    	try {
			    		URL url = new URL(imageUrl);
			    		HttpURLConnection connection  = (HttpURLConnection) url.openConnection();
			    		is = connection.getInputStream();
						os = new FileOutputStream(destFile);
						IOUtils.copy(is, os);
			    	} catch (Exception e) {
			    		setError(e);
			    	} finally {
						if (is != null) try { is.close(); } catch (Exception e) {}
						if (os != null) try { os.close(); } catch (Exception e) {}
			    	}
					return null;
				}
				
				@Override
				protected void onPostExecute(Void result) {
					super.onPostExecute(result);
					if (getError() != null) {
						return;
					}
				    String title = getIntent().getExtras().getString(IMAGE_TITLE);
				    if (title == null) {
				    	title = "";
				    } else {
				    	title = Html.fromHtml(title).toString();
				    }

					ShareUtils.shareImage(ImageViewerActivity.this,
							destFile.getAbsolutePath(),
							"image/jpeg",
							title,
							getString(R.string.share_image_title));
				}
			}.execute();
		} catch (Exception e) {
			DialogUtils.showErrorDialog(this, e);
		}
		
	}

	private void setWallpaper() {
		new AbsProgressBarAsyncTask<Void, Void, Bitmap>(this, getString(R.string.downloading_image)) {
			@Override
			protected Bitmap doInBackground(Void... params) {
				try {
					final String imageUrl = getIntent().getExtras().getString(IMAGE_URL);
					return ImageUtils.readImage(imageUrl);
				} catch (Exception e) {
					setError(e);
				}
				
				return null;
			}
			
			@Override
			protected void onPostExecute(Bitmap bitmap) {
				super.onPostExecute(null);
				if (bitmap == null) {
					return;
				}
				try {
					WallpaperManager wpm = WallpaperManager.getInstance(ImageViewerActivity.this);
		            float width =  wpm.getDesiredMinimumWidth();
					float height =  wpm.getDesiredMinimumHeight();
					bitmap = ImageUtils.getResizedBitmap(bitmap, width, height);
					wpm.setBitmap(bitmap);
		            Toast.makeText(ImageViewerActivity.this,
		            		getString(R.string.wallpaper_changed_title),
							Toast.LENGTH_LONG)
							.show();
				} catch (IOException e) {
					DialogUtils.showErrorDialog(ImageViewerActivity.this, e);
				}
			}
		}.execute();
	}
}

File: imageviewer/src/main/java/com/ternaryop/photoshelf/imageviewer/activity/ImageViewerActivity.kt
package com.ternaryop.photoshelf.imageviewer.activity

import android.annotation.SuppressLint
import android.app.ActivityOptions
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import com.ternaryop.compat.content.getActivityInfoCompat
import com.ternaryop.compat.os.getSerializableCompat
import com.ternaryop.photoshelf.activity.AbsPhotoShelfActivity
import com.ternaryop.photoshelf.activity.ImageViewerData
import com.ternaryop.photoshelf.imageviewer.R
import com.ternaryop.photoshelf.imageviewer.service.WallpaperService
import com.ternaryop.photoshelf.imageviewer.util.ImageViewerUtil
import com.ternaryop.photoshelf.util.menu.enableAll
import com.ternaryop.utils.dialog.showErrorDialog
import com.ternaryop.utils.intent.ShareChooserParams
import com.ternaryop.utils.text.fromHtml
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import java.util.Locale

const val DIMENSIONS_POST_DELAY_MILLIS = 3000L
const val FILE_PROVIDER_SHARE_AUTHORITY = "com.ternaryop.photoshelf.imagepicker.viewer.fileProviderShareAuthority"

private const val SUBDIRECTORY_PICTURES = "TernaryOpPhotoShelf"

@SuppressLint("SetJavaScriptEnabled")
class ImageViewerActivity : AbsPhotoShelfActivity() {
    private var webViewLoaded: Boolean = false
    private var imageHostUrl: String? = null
    private lateinit var detailsText: TextView

    override val contentViewLayoutId: Int = R.layout.activity_webview

    // no frame inside the layout so no need for contentFrameId
    override val contentFrameId: Int = 0

    private lateinit var imageViewerData: ImageViewerData

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        title = ""

        val progressBar = findViewById<View>(R.id.webview_progressbar) as ProgressBar
        detailsText = findViewById<View>(R.id.details_text) as TextView

        imageViewerData = intent.extras?.getSerializableCompat(EXTRA_IMAGE_VIEWER_DATA, ImageViewerData::class.java)
            ?: return

        val data = "<body><img src=\"${imageViewerData.imageUrl}\"/></body>"
        prepareWebView(progressBar).loadDataWithBaseURL(null, data, "text/html", "UTF-8", null)
        try {
            imageHostUrl = URI(imageViewerData.imageUrl).host
        } catch (ignored: URISyntaxException) {
        }
        drawerToolbar.overflowIcon = ContextCompat.getDrawable(this, R.drawable.ic_imageviewer_overflow)
    }

    override fun createFragment(): Fragment? = null

    private fun prepareWebView(progressBar: ProgressBar): WebView {
        val webView = findViewById<View>(R.id.webview_view) as WebView
        webView.settings.javaScriptEnabled = true
        webView.addJavascriptInterface(this, "dimRetriever")
        webViewLoaded = false

        webView.setInitialScale(1)
        webView.settings.loadWithOverviewMode = true
        webView.settings.useWideViewPort = true
        webView.settings.builtInZoomControls = true
        webView.settings.setSupportZoom(true)
        webView.settings.displayZoomControls = false

        webView.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView, newProgress: Int) {
                progressBar.progress = newProgress
            }
        }
        webView.webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
                progressBar.progress = 0
                progressBar.visibility = View.VISIBLE
            }

            override fun onPageFinished(view: WebView, url: String) {
                webViewLoaded = true
                progressBar.visibility = View.GONE
                @Suppress("MaxLineLength")
                view.loadUrl("javascript:var img = document.querySelector('img');dimRetriever.setDimensions(img.width, img.height)")
                invalidateOptionsMenu()
            }
        }
        return webView
    }

    @JavascriptInterface
    @Suppress("unused")
    fun setDimensions(w: Int, h: Int) {
        runOnUiThread {
            detailsText.visibility = View.VISIBLE
            detailsText.text = String.format(Locale.US, "%s (%1dx%2d)", imageHostUrl, w, h)
            Handler(Looper.getMainLooper())
                .postDelayed({ detailsText.visibility = View.GONE }, DIMENSIONS_POST_DELAY_MILLIS)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.image_viewer, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        menu.enableAll(webViewLoaded)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_image_viewer_wallpaper -> {
                WallpaperService.startChange(this, imageViewerData.imageUrl)
                return true
            }
            R.id.action_image_viewer_share -> {
                startShareImage()
                return true
            }
            R.id.action_image_viewer_download -> {
                startDownload()
                return true
            }
            R.id.action_image_viewer_copy_url -> {
                ImageViewerUtil.copyToClipboard(
                    this, imageViewerData.imageUrl,
                    getString(R.string.image_url_description), R.string.url_copied_to_clipboard_title
                )
                return true
            }
            R.id.action_image_viewer_details -> {
                toggleDetails()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    private fun startDownload() {
        val fileName = ImageViewerUtil.buildFileName(imageViewerData.imageUrl, imageViewerData.tag)
        launch {
            try {
                val relativePath = File(SUBDIRECTORY_PICTURES, fileName)
                ImageViewerUtil.download(this@ImageViewerActivity, imageViewerData.imageUrl, relativePath)
            } catch (t: Throwable) {
                t.showErrorDialog(this@ImageViewerActivity)
            }
        }
    }

    private fun startShareImage() {
        try {
            val destFile = ImageViewerUtil.buildSharePath(this, imageViewerData.imageUrl, "images")
            val shareChooserParams = ShareChooserParams(
                FileProvider.getUriForFile(this, getFileProvider(), destFile),
                getString(R.string.share_image_title),
                imageViewerData.title?.fromHtml()?.toString() ?: ""
            )
            launch(Dispatchers.IO) {
                ImageViewerUtil.shareImage(this@ImageViewerActivity, URL(imageViewerData.imageUrl), shareChooserParams)
            }
        } catch (t: Throwable) {
            t.showErrorDialog(this)
        }
    }

    private fun toggleDetails() {
        detailsText.visibility = if (detailsText.visibility == View.GONE) View.VISIBLE else View.GONE
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(com.ternaryop.photoshelf.core.R.anim.slide_in_right, com.ternaryop.photoshelf.core.R.anim.slide_out_right)
    }

    private fun getFileProvider(): String {
        val componentName = ComponentName(this, javaClass)
        val data = packageManager.getActivityInfoCompat(componentName, PackageManager.GET_META_DATA.toLong()).metaData
        return checkNotNull(data.getString(FILE_PROVIDER_SHARE_AUTHORITY)) {
            "Unable to find $FILE_PROVIDER_SHARE_AUTHORITY"
        }
    }

    companion object {
        const val EXTRA_IMAGE_VIEWER_DATA = "com.ternaryop.photoshelf.extra.IMAGE_VIEWER_DATA"

        fun startImageViewer(context: Context, imageViewerData: ImageViewerData) {
            val intent = Intent(context, ImageViewerActivity::class.java)
            val bundle = Bundle()

            bundle.putSerializable(EXTRA_IMAGE_VIEWER_DATA, imageViewerData)
            intent.putExtras(bundle)

            val animBundle = ActivityOptions.makeCustomAnimation(
                context,
                com.ternaryop.photoshelf.core.R.anim.slide_in_left,
                com.ternaryop.photoshelf.core.R.anim.slide_out_left
            ).toBundle()
            context.startActivity(intent, animBundle)
        }
    }
}
